Add Custom VCC Blueprint (#31)
madewithkoji/docs: src/asciidoc/docs/blueprints/customVCC.adoc

> = Cat Selector VCC

Reviewer: ```suggestion
= Cat selector VCC
```
I was really inconsistent in how I converted these page titles, but I'd like to standardize on sentence case. I have some cleanup to do on the rest of the site :)

:page-slug: custom-vcc-blueprint

> Follow this blueprint to build a template for a custom Koji VCC to give remixes the option to select from a drop down of choices which are retrieved from a third party API.

Reviewer: ```suggestion
Follow this blueprint to build a custom VCC that allows remixers to select from a drop-down list of cat breeds.
You will retrieve the list from a third-party API and style the drop-down to look like a native Koji control.
```

Also, I'd like to create a new section on the Blueprints Overview page called "Build your custom VCC" and add the "Cat Selector VCC" as the first item (rather than putting it with the "templates"). For some reason, I don't see that file in the commit so maybe it's not bundled with this pull request or I'm looking in the wrong place.

You will retrieve the list from a third-party API and style the drop-down to look like a native Koji control.

> image::vote-counter.png[Make a custom VCC on Koji,role="img-overview"]

Reviewer: ```suggestion
image::cat-selector.png[Make a custom VCC on Koji,role="img-overview"]
```
Luis is making us a new image today.

== Building blocks

=== Remix the scaffold

Remix an existing web application on Koji that implements basic elements of your favorite framework.

.React
https://withkoji.com/templates/seane/simple-react-scaffold-remix-[Simple React scaffold]

.Vanilla JS
https://withkoji.com/templates/JamesHole/vanilla-js-scaffold-with-vccs[Vanilla JS scaffold]

=== Install the packages

Install @withkoji/custom-vcc-sdk to enable custom VCC functionality, such as loading and saving the VCC value from Koji.

[source,bash]
----
npm install --save @withkoji/custom-vcc-sdk
----

=== Create the customization files
Create `.koji/customization/settings.json` to define customizable values for the API URL, and the items key and item name key from the API response.

[source,json]
----
include::../_includes/vcc-blueprint-settings.json[tags=!configuration]
----
<1> You will add custom values here in the next step.

Remove the unused customization files.

`colors.json`, `images.json`, and `strings.json`

=== Configure the API
This blueprint implements https://thecatapi.com/[The Cat API] to return a list of cat breeds.
However, you can use any API that returns a list of items to implement a custom selector.

> Set the values for the API in the settings customisation file so your app can access the API.

Reviewer: ```suggestion
Set the values for the API in the `settings.json` customization file so that your application can access the API.
```

`colors.json`, `images.json`, and `strings.json`

=== Configure the API

> For this blueprint we implement https://thecatapi.com/[The Cat API] which can return a list of cat breeds but you can use any API that returns a list of items.

Reviewer: As a style convention, I've decided to use "you" instead of "we" for the narrative instructions.

```suggestion
This blueprint implements https://thecatapi.com/[The Cat API] to return a list of cat breeds.
However, you can use any API that returns a list of items to implement a custom selector.
```

----

Make a call to the API endpoint to retrieve the required values.

.React
[source,javascript]
----
include::../_includes/customVCCReact.js[tags=loadAPI,indent=0]
----

.Vanilla JS
[source,javascript]
----
include::../_includes/customVCCVanilla.js[tags=loadAPI,indent=0]
----

Create a setter to return the selected value from your application back to the VCC on Koji.

Use `customVCC.change()` to notify Koji that the VCC value has been updated and `customVCC.save()` to tell Koji to commit the change to the remixer's VCC.

.React
[source,javascript]
----
include::../_includes/customVCCReact.js[tags=selectValue,indent=0]
----
.Vanilla JS
[source,javascript]
----
include::../_includes/customVCCVanilla.js[tags=selectValue,indent=0]
----

Use the Koji theme and the API results to render a drop-down menu that allows users to select a value.

.React
[source,javascript]
----
include::../_includes/customVCCReact.js[tags=render,indent=0]
----

.Vanilla JS
[source,javascript]
----
include::../_includes/customVCCVanilla.js[tags=render,indent=0]
----

=== Add the styles
Define styles for the template, using the theme values to match the user's Koji theme.

> 

Reviewer: Luis noticed that the Koji logo is covering the selection in the example "template". Can you apply some responsive styling to adjust the vertical spacing?

Author: The Koji logo only shows on a published preview window, which you wouldn't have a reason to view for a VCC? The reason it is styled close to the top is so if you switch to a regular `type:select` VCC they display consistently.

Reviewer: I believe it's better to have a consistent VCC input than a preview that isn't obscured. Let's leave it for now and we can circle back if it raises questions.

.React
[source,javascript]
----
include::../_includes/customVCCReact.js[tags=styles,indent=0]
----

.Vanilla JS
[source,css]
----
include::../_includes/vcc-blueprint-styles.css[]
----

=== Test

> To save development time, Koji provides an easy way to test a custom VCC before it is published. First you will need to create a test consumer project, do so by remixing https://withkoji.com/templates/seane/simple-react-scaffold-remix-[Simple React Scaffold].

Reviewer: ```suggestion
To save development time, Koji provides an easy way to test a custom VCC before it is published.
```
Let's move this info into the procedure. I'll provide a suggestion below.

> . In your custom VCC project, click the Remote tab in the right pane.

Reviewer: ```suggestion
. In your custom VCC project, click the *Remote* tab in the right pane.
```

> . Click Copy URL to get the staging URL for your project.

Reviewer: ```suggestion
. Click *Copy URL* to get the staging URL for your project.
```

> . Open your consumer project.

Reviewer: ```suggestion
. Open another project to act as the "consumer" for testing the VCC.
+
TIP: If you don't have another project, you can create a test project by remixing the
https://withkoji.com/templates/seane/simple-react-scaffold-remix-[Simple React Scaffold].
```

> . In your consumer project, open the Strings customization file, and click Code to view the raw JSON.

Reviewer: ```suggestion
. In your consumer project, open a JSON customization file, and add or edit a field to the custom VCC.
+
The type must match this format, using the URL you just copied as YOUR-URL:
+
`"type": "custom<YOUR-URL>"`
```
I'm trying to make these instructions a bit more generic vs. the tutorial, which is very specific. But, you might need to play around with the steps and language a bit to make it read more clearly when it's all in sequence.

> . For the title field, change the type to match this format, using the URL you just copied as YOUR-URL: +

Reviewer: If you use the suggestion above, it replaces this line and the next.

`"type": "custom<YOUR-URL>"`

> . Save the file, and return to the Visual view of the customization file. You should see your custom VCC.

Reviewer: ```suggestion
. Save the file, and return to the Visual view of the customization file.
You should see your custom VCC.
```
